File: ios/App/Binding/StatsBinding.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2023 Blocka AB. All rights reserved.
//
//  @author Kar
//

import Foundation
import Factory
import Combine

class StatsBinding: StatsOps {
    let blockedCounter = CurrentValueSubject<String, Never>("")

    @Injected(\.flutter) private var flutter

    init() {
        StatsOpsSetup.setUp(binaryMessenger: flutter.getMessenger(), api: self)
    }

    func doBlockedCounterChanged(blocked: String, completion: @escaping (Result<Void, Error>) -> Void) {
        blockedCounter.send(blocked)
        completion(.success(()))
    }
}

extension Container {
    var stats: Factory<StatsBinding> {
        self { StatsBinding() }.singleton
    }
}

File: android/app/src/main/java/binding/StatsBinding.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2023 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package binding

import channel.stats.StatsOps
import service.FlutterService

object StatsBinding: StatsOps {
    private val flutter by lazy { FlutterService }

    var blocked = ""

    init {
        StatsOps.setUp(flutter.engine.dartExecutor.binaryMessenger, this)
    }

    override fun doBlockedCounterChanged(blocked: String, callback: (Result<Unit>) -> Unit) {
        this.blocked = blocked
        callback(Result.success(Unit))
    }
}